tests/instrumentation/asyncio/asyncpg_tests.py

```diff
@@ -0,0 +1,116 @@
+#  BSD 3-Clause License
+#
+#  Copyright (c) 2019, Elasticsearch BV
+#  All rights reserved.
+#
+#  Redistribution and use in source and binary forms, with or without
+#  modification, are permitted provided that the following conditions are met:
+#
+#  * Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+#  * Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+#  * Neither the name of the copyright holder nor the names of its
+#    contributors may be used to endorse or promote products derived from
+#    this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+#  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+#  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+#  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+#  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+#  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+#  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import pytest  # isort:skip
+
+asyncpg = pytest.importorskip("asyncpg")  # isort:skip
+
+import os
+
+from elasticapm.conf import constants
+
+try:
+    from psycopg2 import sql
+
+    has_sql_module = True
+except ImportError:
+    # as of Jan 2018, psycopg2cffi doesn't have this module
+    has_sql_module = False
+
+pytestmark = [pytest.mark.asyncpg, pytest.mark.asyncio]
+
+if "POSTGRES_DB" not in os.environ:
+    pytestmark.append(pytest.mark.skip("Skipping asyncpg tests, no POSTGRES_DB environment variable set"))
+
+
+def dsn():
+    return "postgres://{user}:{password}@{host}:{port}/{database}".format(
+        **{
+            "database": os.environ.get("POSTGRES_DB", "elasticapm_test"),
+            "user": os.environ.get("POSTGRES_USER", "postgres"),
+            "password": os.environ.get("POSTGRES_PASSWORD", "postgres"),
+            "host": os.environ.get("POSTGRES_HOST", "localhost"),
+            "port": os.environ.get("POSTGRES_PORT", "5432"),
+        }
+    )
+
+
+@pytest.fixture()
+async def cursor(request):
+    conn = await asyncpg.connect(dsn())
+    cur = await conn.cursor()
+
+    # we use a finalizer instead of yield, because Python 3.5 throws a syntax error, even if the test doesn't run
+    def rollback():
```

Comment:
This test is excluded from  from Python 3.5 run, would suggest just adding this after `yield cursor` and awaiting the ROLLBACK and close

Reply:
Unfortunately, the code is still parsed in Python 3.5 runs (during pytest's discovery phase), leading to the `SyntaxError` mentioned in the comment.

Once we deprecate Python 3.5, we can change it to the `yield` approach.

Reply:
Looks like this is a problem with `asyncpg` compared to the other postgres drivers. We need to await both the rollback and the `conn.close()`. Which is a problem, as @beniwohli mentioned before, because python3.5 throws a syntax error.

Python 3.5 is EOL one month from today. We may just wait to finish this PR until that deprecation takes place. Alternatively, because it's so soon, we may jump the gun a little and start that deprecation sooner rather than later so we can get this merged.

Reply:
thank you for a thorough explanation @basepi and @beniwohli! would it be possible to go ahead and merge this PR given that the implementation itself is correct and the tests pass/would be expected to pass, as the alternative way mentioned by @basepi? if yes, what would I have to change/add/remove? as the features in this PR are rather highly expected on our side.

Reply:
I think if @beniwohli is alright with it, I'm going to remove python 3.5 from our matrix tests later this week when I have some time. In the meantime, can you change this test to properly `await` both the rollback and the close as @rafalstapinski recommended? Then when I remove python3.5 the tests should start passing and we can get this merged.

Reply:
done